healthcheck: add regular chain backend and disk space checks

carlaKC opened the pull request:
This PR wraps our existing logger in one which will request graceful shutdown when the `Critical` logging level is used. This wrapper is used to add two health checks to lnd which will shutdown if we do not succeed after sufficient backoff/retry:
1. Check that we have a connection to our chain backend
2. Check that we have some space available on disk

The regularity and strictness (ie, number of retries we allow) of this check are made configurable, so that users can turn them off/ make them less strict, or turn them off entirely if they cause issues (particularly thinking about neutrino flakes).

Windows is windows, so we need to have a special windows space check (`_windows`, TIL!). 

Partially addresses #4306 (does not cover failed disk writes, so issue not entirely fixed)

Fixes #1557
Fixes #3109

Roasbeef commented on build/log_shutdown.go:
```diff
@@ -0,0 +1,59 @@
+// function to prompt safe shutdown.
+//
+// Note: is it part of the btclog.Logger interface.
+func (s *ShutdownLogger) Criticalf(format string, params ...interface{}) {
```
👍

Roasbeef commented on build/log_shutdown.go:
```diff
@@ -0,0 +1,59 @@
+// use the signal package to request shutdown on critical errors.
+func NewShutdownLogger(logger btclog.Logger) *ShutdownLogger {
+	shutdown := func() {
+		// If our application is not listening for interrupts, we return
```
Can't this just be a regular method body given it closes over the logger which is already embedded as an attribute in the struct?

Roasbeef commented on signal/signal.go:
```diff
@@ -45,6 +48,10 @@ func Intercept() {
 func mainInterruptHandler() {
+	// Close our started channel to signal that we have started listening
+	// for interrupts.
+	close(started)
```
This would cause a double close panic if `Intercept` is ever called twice....

Perhaps we should try to enforce that there is only a singleton registration for interrupt signals?

Roasbeef commented on config.go:
```diff
@@ -86,6 +86,10 @@ const (
 	// backing IP of a host has changed.
 	defaultHostSampleInterval = time.Minute * 5
+
+	defaultChainInterval = time.Minute * 10
```
Hmm, this seems a bit too high...I'd imagine we'd want to check at least once a minute, just so any hypervisor-like monitor can restart `lnd` shortly after the backend shuts down.

Roasbeef replied:
An alternative to polling would be to rely on the `zmq` connection for `bitcoind` and the websockets connection for `btcd`. If either of these connections fail (EOF error), then we know the backend is down.

carlaKC replied:
Relying on these connections meaning make a change in `btcwallet` so that we can access the existing connection, or add a direct connection to zmq/websockets from the health check?

Roasbeef replied:
Thinking just a direction connection so we don't need to dig down into the wallet. For `bitcoind`, we can only request the block notifications which aren't as spammy. A similar option can be applied to the `btcd` web sockets connection.

Roasbeef commented on healthcheck/healthcheck.go:
```diff
@@ -0,0 +1,167 @@
+// Package healthcheck contains takes contains a monitor which takes a set
```
👍

Roasbeef commented on healthcheck/healthcheck.go:
```diff
@@ -0,0 +1,167 @@
+func retryCheck(check *Observation, quit chan struct{}, shutdown shutdownFunc) {
+	var count int
+	for count < check.Retries {
+		err := check.Check()
```
if retries is zero (lnd should shutdown if the check fails even once), then this loop is never executed. We want more of a "do -- while" loop here.

Roasbeef commented on healthcheck/healthcheck_test.go:
```diff
@@ -0,0 +1,195 @@
+		{
+			name:             "first call succeeds",
+			callErrors:       []error{nil},
+			retries:          2,
```
Should add a test case for zero retries.

Roasbeef commented on server.go:
```diff
@@ -1240,6 +1244,34 @@ func newServer(cfg *Config, listenAddrs []net.Addr, chanDB *channeldb.DB,
+	var healthChecks []*healthcheck.Observation
+	if !cfg.HealthChecks.ChainCheck.Disabled {
+		backendCheck := func() error {
+			_, _, err := cc.chainIO.GetBestBlock()
```
Need to double check how this degrades for each backend...I think for `btcd` this might just block indefinitely rather than actually returning immediately with an error.

carlaKC commented:
Rebased and updated in separate pushes, diff of changes [here](https://github.com/lightningnetwork/lnd/compare/42ee2696779367db67ad554b87badd86d9281295..e776d99ec1277fbe8a6df0797119583b7d00b9fd). Changes since last review:
- Update changes to signal package so that we can only call intercept once
- Add timeout to health checks (because btcd `GetBestBlock` blocks when btcd shuts down)
- Add disk check for solaris behind build tag, as done for windows

Crypt-iQ commented on build/log_shutdown.go:
```diff
@@ -0,0 +1,53 @@
+// and writes to log with LevelCritical. It will then call the shutdown
+// logger's shutdown function to prompt safe shutdown.
+//
+// Note: is it part of the btclog.Logger interface.
```
nit: It is

Crypt-iQ commented on healthcheck/diskcheck_solaris.go:
```diff
@@ -0,0 +1,17 @@
+
+	// Calculate our free blocks/total blocks to get our
+	// total number of free blocks.
+	return float64(s.Blocks) / float64(s.Blocks), nil
```
`s.Blocks` -> `s.Bfree` ?

Roasbeef commented:
> Is it possible to make integration tests out of these with the various backends and kill the backends to see that lnd does shutdown? I don't think it would be possible to really test the disk healthcheck. 

That's possible, but the test set up would need to be slightly distinct as all the other tests assumes that the target chain backend will be around the entire time. Agreed about the disk check....one way to test it would be using a container w/ a very small volume (would basically be unable to start).

Crypt-iQ commented on lncfg/healthcheck.go:
```diff
@@ -23,11 +24,27 @@ var (
+	if h.DiskCheck.RequiredRemaining < 0 ||
+		h.DiskCheck.RequiredRemaining >= 1 {
+
+		return errors.New("disk required ratio must be in [0;1)")
```
`[0:1)` maybe?

carlaKC commented:
> Haven't been able to test this out

I have run some local tests for disk check, and chain backend for btcd + bitcoind, so it depends on how much time we want to put into testing. I can wrangle together some bash scripts to put up and kill docker containers if needed. 

> `btcd` hang comment above

Right now `btcd` does hang, which is why the timeout was added for these checks. The unfortunate part is that the goroutine calling the best block endpoint will hang indefinitely, but there's not much that can be done unless we make a change to btcwallet. There is the option of using zmq/websockets but that comes with some copy-fu from btcwallet, and halves the number of lnd nodes bitcoind can serve (down to 2, which sucks for local dev).

Crypt-iQ commented:
> > Haven't been able to test this out
> 
> I have run some local tests for disk check, and chain backend for btcd + bitcoind, so it depends on how much time we want to put into testing. I can wrangle together some bash scripts to put up and kill docker containers if needed.
> 
> > `btcd` hang comment above
> 
> Right now `btcd` does hang, which is why the timeout was added for these checks. The unfortunate part is that the goroutine calling the best block endpoint will hang indefinitely, but there's not much that can be done unless we make a change to btcwallet. There is the option of using zmq/websockets but that comes with some copy-fu from btcwallet, and halves the number of lnd nodes bitcoind can serve (down to 2, which sucks for local dev).

Would be super helpful to have some docker scripts, maybe could even be run as part of our tests?
